ccvm_simulators/metadata.py

```diff
@@ -45,9 +52,12 @@ def save_metadata_to_file(self, file_dir="./metadata", file_name="metadata"):
 
         metadata_file_path = f"{file_dir}/{file_name}.json"
 
+        if self.metadata_dict is None:
+            self.finalize_metadata()
+
```

Comment:
The current implementation updates the `self.metadata_dict` only when it is `None`. But, there can be an edge case where `def save_metadata_to_file()` is called, and then more result_dict objects could be added to the list. And then if `def save_metadata_to_file()` is called again, `self.metadata_dict` will not be updated again because it is not `None` anymore. I know the chance for such a case is pretty low, but I believe it is still possible to happen.

What I suggest is to just directly update the `self.metadata_dict` whenever a result_dict object is added. By doing so, we would not need to have separate `self.device` and `self.result_metadata` as attributes. Furthermore, we do not need to have `def finalize_metadata()` and when `def save_metadata_to_file()` is called, we can return the `self.metadata_dict` immediately without any intermediate steps.

Reply:
Great suggestion. Done.